src/ProtoBuf.FSharp/ProtobufUtils.fs

```diff
@@ -16,25 +17,27 @@ module Serialiser =
         model.Add(tp, false).SetSurrogate surrogateType
         surrogateType
 
-    /// The magic number where if a union type has more than the above cases it simply is a tagged instance of the parent type.
-    /// Otherwise for this number and below even non-empty unions get their own inner class prefixed with "_".
-    let [<Literal>] private CasesCountWhereNoFieldCasesGenerateType = 3
 
-    /// Allows users to register option types in advance. You can specify a custom suffix name for the Protobuf wrapper type generated.
-    /// This only needs to be called directly if your type either is not already a field in another type previously registered (e.g. a record or union)
-    /// and/or your not happy with the default type name in case of naming clashes.
-    /// By default if None is provided for the customTypeSuffix parameter for example with Option<string> the protobuf message will be an "OptionalString".
-    /// If the model is already registered (explictly or implicitly via another registration) AND/OR the type passed in is not an option type this will no-op.
-    let registerOptionTypesIntoModel (optionType: Type) customTypeSuffix (model: RuntimeTypeModel) =
-        if optionType.IsGenericType && optionType.GetGenericTypeDefinition() = typedefof<Option<_>> then
-            let definedTypes = seq {
-                for m in model.GetTypes() do
-                let m = m :?> MetaType
-                yield m.Type
-            }
-            if definedTypes |> Seq.contains optionType |> not
-            then
-                registerSurrogate optionType model |> ignore
+    let internal collectionTypes =
+        HashSet (seq { typedefof<_ list>; typedefof<Set<_>>; typedefof<Map<_, _>> })
+
+    let private internalRegisterCollection (listType: Type) (model: RuntimeTypeModel) =
+        if model.IsDefined(listType) |> not then
+            let t = model.Add(listType, false)
+            t.IgnoreListHandling <- true
+            CodeGen.getSurrogate listType |> t.SetSurrogate

```

Comment:
Nice that this seems to work now.

Reply:
IgnoreListHandling does. model.IsDefined still behaves weird --- it didn't cause problems here, but when I try to use it on option types it causes tests to fail. Keeping separate set of registered F# types for model is probably still a good idea.